src/engine/UNSATCertificate.h

```diff
@@ -0,0 +1,282 @@
+    void wasVisited();
+
+    /*
+      Sets value of _shouldDelegate to be true
+      Saves delegation to file iff saveToFile is true
+    */
+    void shouldDelegate( unsigned delegationNumber, DelegationStatus saveToFile );
+
+    /*
+     Removes all PLC explanations
+    */
+    void deletePLCExplanations();
+
+    /*
+     Removes all PLC explanations from a certain point
+    */
+    void resizePLCExplanationsList( unsigned newSize );
+
+    /*
+      Deletes all offsprings of the node and makes it a leaf
+    */
+    void makeLeaf();
+
+    /*
+      Removes all PLCExplanations above a certain decision level WITHOUT deleting them
+      ASSUMPTION - explanations pointers are kept elsewhere before removal
+    */
+    void removePLCExplanationsBelowDecisionLevel( unsigned decisionLevel );
+
+private:
+    List<CertificateNode*> _children;
+    List<ProblemConstraint> _problemConstraints;
+    CertificateNode* _parent;
+    List<PLCExplanation*> _PLCExplanations;
+    Contradiction *_contradiction;
+    PiecewiseLinearCaseSplit _headSplit;
+
+    // Enables certifying correctness of UNSAT leaves in SAT queries
+    bool _hasSATSolution;
+    bool _wasVisited;
+
+    DelegationStatus _delegationStatus;
+    unsigned _delegationNumber;
+
+    Vector<Vector<double>> *_initialTableau;
+    Vector<double> _groundUpperBounds;
+    Vector<double> _groundLowerBounds;
+
+    /*
+      Copies initial tableau and ground bounds
+    */
+    void copyGroundBounds( Vector<double> &groundUpperBounds, Vector<double> &groundLowerBounds );
+
+    /*
+      Inherits the initialTableau pointer, the ground bounds and the problem constraint from parent, if exists.
+      Fixes the phase of the constraint that corresponds to the head split
+    */
+    void passChangesToChildren( ProblemConstraint *childrenSplitConstraint );
+
+    /*
+      Checks if the node is a valid leaf
+    */
+    bool isValidLeaf() const;
+
+    /*
+      Checks if the node is a valid none-leaf
+    */
+    bool isValidNoneLeaf() const;
+
+    /*
+      Write a leaf marked to delegate to a smtlib file format
+    */
+    void writeLeafToFile();
+
+    /*
+      Return true iff a list of splits represents a splits over a single variable
+    */
+    bool certifySingleVarSplits( const List<PiecewiseLinearCaseSplit> &splits ) const;
+
+    /*
+      Return true iff the changes in the ground bounds are certified, with tolerance to errors with at most size epsilon
+    */
+    bool certifyAllPLCExplanations( double epsilon );
+
+    /*
+      Return a pointer to the problem constraint representing the split
+    */
+    ProblemConstraint *getCorrespondingReLUConstraint( const List<PiecewiseLinearCaseSplit> &splits );
+
+    /*
+      Certifies a contradiction
+    */
+    bool certifyContradiction();
+
+    /*
+      Computes a bound according to an explanation
+    */
+    double explainBound( unsigned var, bool isUpper, Vector<double> &explanation );
+};
+
+class UNSATCertificateUtils
```

Comment:
This file has too many classes. Maybe we should open a "proofs" folder, in parallel to the engine folder?

Reply:
It makes sense. In this case will there be a file for each class and sturct?